### PR TITLE
Send smart aliases via gRPC to engine

### PR DESCRIPTION
**Pull request by Frassle**
# Description

This add an new field on the RegisterResource and ReadResource protobuf messages to the resource monitor. As well as the `repeated string aliases` field, we now have a `repeated Alias smartAliases` field (name open to scrutiny). The `Alias` object is a protobuf encoding to match the alias objects we have in the SDKs where users can just set "name", or "type". Those objects worked by the SDK then resolving down to a URN, that resolution logic is now also in the engine and so eventually we'll be able to remove all that logic from every SDK (we can't remove it straight away because of back compatibility, a user may have a new SDK but old engine).

Fixes https://github.com/pulumi/pulumi/issues/9734

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by AaronFriel** [comment]
Curious how big of a change in the engine would it be to rename the existing field (keeping tag ID the same) to `alias_urns` and naming the new field `aliases`?

I'd then add a `[deprecated = true]` to the old field, which is a no-op in most languages but might add an annotation or IDE hint.

**Comment by Frassle** [comment]
> Curious how big of a change in the engine would it be to rename the existing field (keeping tag ID the same) to alias_urns and naming the new field aliases? I believe protobufs allow these sort of renames as the thing that matters are the type and tag numbers?
> 
> I'd then add a [deprecated = true] to the old field, which is a no-op in most languages but might add an annotation or IDE hint.

Done. Interestingly a discovery fell out of that rename that we didn't look at aliases for resource reads, and only the Go SDK tried to send them. I'm not sure if aliases make sense for reads anyway so I've just removed that field for now.

Would be good to get @pgavlin check on if alias for reads is/isn't needed.

File: pkg/resource/deploy/deploytest/resourcemonitor.go [hunk 1]
```diff
@@ -104,6 +104,7 @@ type ResourceOptions struct {
 	Remote                  bool
 	Providers               map[string]string
 	AdditionalSecretOutputs []resource.PropertyKey
+	SmartAliases            []resource.Alias
```
Comment (AaronFriel) [inline pkg/resource/deploy/deploytest/resourcemonitor.go:107]: Able to rename this field as well?

**Review by AaronFriel — APPROVED** [review]
My only comment is that this makes me long for codegen-ing resource options. It's a lot of little changes to review in each SDK.

Looking forward to deleting "inheritedChildAlias" and all the other machinery in the SDKs in 4.0.

**Comment by Frassle** [comment]
> My only comment is that this makes me long for codegen-ing resource options. It's a lot of little changes to review in each SDK.

Working on a design for that today 😄 

> Able to rename this field as well?

Yes will do!

**Comment by AaronFriel** [comment]
LGTM

**Pull request merged**